control/controlclient/direct.go

```diff
@@ -950,29 +968,25 @@ func encode(v interface{}, serverKey key.MachinePublic, mkey key.MachinePrivate)
 	return mkey.SealTo(serverKey, b), nil
 }
 
-func loadServerKey(ctx context.Context, httpc *http.Client, serverURL string) (key.MachinePublic, error) {
-	req, err := http.NewRequest("GET", serverURL+"/key", nil)
+func loadServerPubKeys(ctx context.Context, httpc *http.Client, serverURL string) (*tailcfg.OverTLSPublicKeyResponse, error) {
+	keyURL := fmt.Sprintf("%v/key?v=%d", serverURL, tailcfg.CurrentCapabilityVersion)
+	req, err := http.NewRequestWithContext(ctx, "GET", keyURL, nil)
 	if err != nil {
-		return key.MachinePublic{}, fmt.Errorf("create control key request: %v", err)
+		return nil, fmt.Errorf("create control key request: %v", err)
 	}
-	req = req.WithContext(ctx)
 	res, err := httpc.Do(req)
 	if err != nil {
-		return key.MachinePublic{}, fmt.Errorf("fetch control key: %v", err)
+		return nil, fmt.Errorf("fetch control key: %v", err)
 	}
 	defer res.Body.Close()
-	b, err := ioutil.ReadAll(io.LimitReader(res.Body, 1<<16))
-	if err != nil {
-		return key.MachinePublic{}, fmt.Errorf("fetch control key response: %v", err)
-	}
 	if res.StatusCode != 200 {
-		return key.MachinePublic{}, fmt.Errorf("fetch control key: %d: %s", res.StatusCode, string(b))
+		return nil, fmt.Errorf("fetch control key: %d", res.StatusCode)
 	}
-	k, err := key.ParseMachinePublicUntyped(mem.B(b))
-	if err != nil {
-		return key.MachinePublic{}, fmt.Errorf("fetch control key: %v", err)
+	var out tailcfg.OverTLSPublicKeyResponse
+	if err := json.NewDecoder(res.Body).Decode(&out); err != nil {
```

Comment:
You need to be prepared for an old control server here that doesn't return JSON even with the `?v=N` param. So you need to read the res.Body into memory and first try JSON, else treat it as the legacy pub key (without the `mkey:` prefix).

Reply:
Good catch, done.